contracts/EtomicSwap.sol

```diff
@@ -115,25 +200,194 @@ contract EtomicSwap {
+                _amount
+            )
+        );
+
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
+        payments[_id].state = PaymentState.SenderRefunded;
+
+        IERC1155 token = IERC1155(_tokenAddress);
+        token.safeTransferFrom(
+            address(this),
+            msg.sender,
+            _tokenId,
+            _amount,
+            ""
+        );
+
+        emit SenderRefunded(_id);
+    }
+
+    function onERC1155Received(
+        address, /* operator */
+        address from,
+        uint256 tokenId,
+        uint256 value,
+        bytes calldata data
+    ) external override returns (bytes4) {
+        // Decode the data to extract HTLC parameters
+        (
+            bytes32 id,
+            address receiver,
+            address tokenAddress,
+            bytes20 secretHash,
+            uint64 lockTime
+        ) = abi.decode(data, (bytes32, address, address, bytes20, uint64));
+
+        require(
+            receiver != address(0) &&
+            tokenAddress != address(0) &&
+            value > 0 &&
+            payments[id].state == PaymentState.Uninitialized
+        );
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                receiver,
+                from,
+                secretHash,
+                tokenAddress,
+                tokenId,
+                value
+            )
+        );
+
+        payments[id] = Payment(paymentHash, lockTime, PaymentState.PaymentSent);
+        emit PaymentSent(id);
+
+        // Return this magic value to confirm receipt of ERC1155 token
+        return this.onERC1155Received.selector;
+    }
+
+    function onERC1155BatchReceived(
+        address, /* operator */
+        address, /* from */
+        uint256[] calldata, /* ids */
+        uint256[] calldata, /* values */
+        bytes calldata /* data */
+    ) external pure override returns (bytes4) {
+        revert("Batch transfers not supported");
+    }
+
+    function supportsInterface(bytes4 interfaceId)
+    public
+    view
+    override(ERC165, IERC165)
+    returns (bool)
+    {
+        return
+            interfaceId == type(IERC1155Receiver).interfaceId ||
+            super.supportsInterface(interfaceId);
+    }
+
+    function onERC721Received(
+        address, /* operator */
+        address from,
+        uint256 tokenId,
+        bytes calldata data
+    ) external override returns (bytes4) {
+        // Decode the data to extract HTLC parameters
+        (
+            bytes32 id,
+            address receiver,
+            address tokenAddress,
+            bytes20 secretHash,
+            uint64 lockTime
+        ) = abi.decode(data, (bytes32, address, address, bytes20, uint64));
+
+        require(
+            receiver != address(0) &&
+            tokenAddress != address(0) &&
+            payments[id].state == PaymentState.Uninitialized
+        );
```

Comment:
I think it's worth adding the following conditions:
1. `msg.sender == tokenAddress` to ensure that the function is called by the token contract. Maybe we can even reduce the data size and use `msg.sender` instead of `tokenAddress`. I'm not sure regarding the possible security risks of this, please check that it's safe :slightly_smiling_face: 
1. `tx.origin == from && tx.origin == operator` to ensure that token transfer is made by user account itself.
1. Might be also worth checking that receiver is not a smart contract address.

Reply:
1. `msg.sender` usage instead of `tokenAddress` is generally safe in the context that NFT implement erc721/erc1155 standard behavior. But to mitigate the risk of swap of the ERC721/ERC1155 tokens having additional custom behavior I would prefer to leave `msg.sender == tokenAddress` check. 

2. The use of `tx.origin` is discouraged in smart contracts due to [security concerns](https://neptunemutual.com/blog/issues-with-authorization-using-txorigin/). This can be vulnerable to phishing-like attacks, lets not use it. Well we still can do `operator == from` check to ensure that operator performing the transfer is the same as the current NFT owner.

3. why not, added this check.

Reply:
Nice, thanks 🙂